samples/bluetooth/hci_uart_async/src/hci_uart_async.c

```diff
@@ -0,0 +1,371 @@
+	}
+
+	LOG_DBG("Pattern found");
+	h2c_cmd_reset = bt_buf_get_tx(BT_BUF_H4, K_FOREVER, h4_cmd_reset, sizeof(h4_cmd_reset));
+	LOG_DBG("Fowarding reset");
+
+	err = bt_send(h2c_cmd_reset);
+	__ASSERT(!err, "Failed to send reset: %d", err);
+}
+
+static void h2c_h4_transport(void)
+{
+	for (;;) {
+		int err;
+		struct net_buf *buf;
+		uint8_t h4_type;
+		uint8_t hdr_size;
+		uint8_t *hdr_buf;
+		uint16_t payload_size;
+
+
+		LOG_DBG("h2c: listening");
+
+		/* Read H4 type. */
+		err = uart_h2c_rx(&h4_type, sizeof(uint8_t));
+
+		if (err) {
+			return;
+		}
+		LOG_DBG("h2c: h4_type %d", h4_type);
+
+		/* Allocate buf. */
+		buf = bt_buf_get_tx(BT_BUF_H4, K_FOREVER, &h4_type, sizeof(h4_type));
+		LOG_DBG("h2c: buf %p", buf);
+
+		if (!buf) {
+			/* `h4_type` was invalid. */
+			__ASSERT_NO_MSG(hci_hdr_size(h4_type) == 0);
+
+			LOG_WRN("bt_buf_get_tx failed h4_type %d", h4_type);
+			return;
+		}
+
+		/* Read HCI header. */
+		hdr_size = hci_hdr_size(h4_type);
+		hdr_buf = net_buf_add(buf, hdr_size);
+
+		err = uart_h2c_rx(hdr_buf, hdr_size);
+		if (err) {
+			net_buf_unref(buf);
+			return;
+		}
+		LOG_HEXDUMP_DBG(hdr_buf, hdr_size, "h2c: hci hdr");
+
+		/* Read HCI payload. */
+		payload_size = hci_payload_size(hdr_buf, h4_type);
+
+		LOG_DBG("h2c: payload_size %u", payload_size);
+
+		if (payload_size <= net_buf_tailroom(buf)) {
+			uint8_t *payload_dst = net_buf_add(buf, payload_size);
+
+			err = uart_h2c_rx(payload_dst, payload_size);
+			if (err) {
+				net_buf_unref(buf);
+				return;
+			}
+			LOG_HEXDUMP_DBG(payload_dst, payload_size, "h2c: hci payload");
+		} else {
+			uint8_t *discard_dst;
+			uint16_t discard_size;
+
+			/* Discard oversize packet. */
+			LOG_WRN("h2c: Discarding oversize h4_type %d payload_size %d.", h4_type,
+				payload_size);
+
+			net_buf_reset(buf);
+			discard_dst = net_buf_tail(buf);
+			discard_size = net_buf_max_len(buf);
+
+			while (payload_size) {
+				uint16_t read_size = MIN(payload_size, discard_size);
+
+				err = uart_h2c_rx(discard_dst, read_size);
+				if (err) {
+					net_buf_unref(buf);
+					return;
+				}
+
+				payload_size -= read_size;
+			}
+
+			net_buf_unref(buf);
+			buf = NULL;
+		}
+
+		LOG_DBG("h2c: packet done");
+
+		/* Route buf to Controller. */
+		if (buf) {
+			err = bt_send(buf);
+			if (err) {
+				/* This is not a transport error. */
+				LOG_ERR("bt_send err %d", err);
+				net_buf_unref(buf);
+				buf = NULL;
+			}
+		}
+
+		k_yield();
+	}
+}
+
+static void h2c_thread_entry(void *p1, void *p2, void *p3)
+{
+	k_thread_name_set(k_current_get(), "HCI TX (h2c)");
+
+	for (;;) {
+		LOG_DBG("Synchronized");
+		h2c_h4_transport();
+		LOG_WRN("Desynchronized");
+		send_hw_error();
+		recover_sync_by_reset_pattern();
+	}
+}
+
+void callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	if (evt->type == UART_RX_DISABLED) {
+		k_poll_signal_raise(&uart_h2c_rx_sig, 0);
+	}
+	if (evt->type == UART_RX_STOPPED) {
```

Comment:
you can use `else` since only one event arrives at time.

Minor optimization: what about setting signal variable and variable in the `if` and then calling `k_poll_signal_raise` only once (should be less code).